colorSV (<ins>co</ins>assembly <ins>lo</ins>ng-<ins>r</ins>ange <ins>SV</ins> caller) identifies long-range somatic structural variation by examining the local topology of matched tumor-normal co-assembly graphs. The input is a single joint assembly graph in `.gfa` format (generated by an assembler such as [hifiasm](https://github.com/chhylp123/hifiasm)), and the output is a set of SV calls.

# Table of Contents
* [Software Requirements](#software-requirements)
* [Example Usage](#example-usage)
* [Pipeline](#pipeline)
	* [1) Joint Assembly](#joint-assembly)
	* [2) Preprocessing](#preprocessing)
	* [3) SV Calling](#sv-calling)


# Software Requirements
* [Minimap2](https://github.com/lh3/minimap2), v2.27 or later
* [k8](https://github.com/attractivechaos/k8), v1.0 or later


# Example Usage

```
# Perform co-assembly
hifiasm -o demo.asm ./example/demo_samples.fa 2> demo.log

# Identify and align tumor-only unitigs
color-sv preprocess -o ./demo_results --graph demo.asm.bp.r_utg.gfa --reference mini_ref.fa --tumor-ids tumor_sample1,tumor_sample2 --read-sep /

# Perform SV calling
color-sv call -o ./demo_results --graph demo.asm.bp.r_utg.gfa --filter ./example/mask_regions.bed
```

# Pipeline
The colorSV method performs SV calling in three steps: 1) generation of a co-assembly graph using an external assembler, 2) identification and alignment of tumor-only graph nodes, and 3) filtering of candidate breakpoints by performing a local topology search on the co-assembly graph. The second and third steps are separated so that the SV calling step can be performed on the same graph multiple times (e.g., using different parameters) without having to rerun tumor-only node identification and sequence alignment.

## 1) Joint Assembly
colorSV requires a matched tumor-normal co-assembly graph as input. This graph should be generated by pooling reads from matched tumor and normal samples, then combining them into a single assembly. For instance, assembling the reads with [hifiasm](https://github.com/chhylp123/hifiasm) will generate multiple co-assembly graphs in the .gfa format. We would recommend using the raw unitig graph (with the .asm.bp.r\_utg.gfa suffix), as using the processed unitig or contig graphs (with the p\_utg or p\_ctg suffixes) may reduce sensitivity due to the removal of small bubbles.

## 2) Preprocessing
The next step of the pipeline identifies and aligns tumor-only nodes of the co-assembly graph (nodes that are only supported by reads from the tumor sample) to the reference genome. The command can be run with

```
./color-sv preprocess -o /path/to/output/directory/ --graph coassembly_graph.gfa --reference ref.fa --tumor-ids id1,id2 --read-sep <sep> [optional flags]
```

This command will create the directory `/path/to/output/directory/intermediate_output/`, which will be used to store the IDs and alignments of the tumor-only nodes. More information about the options:

* Required arguments
	* `--graph`: path to a joint assembly graph in .gfa format (see the section on [joint assembly](#joint-assembly))
	* `--reference`: path to the reference genome in .fa format (we would recommend using T2T-CHM13v2.0)
	* `--tumor-ids`: a list of the tumor sample IDs (`--tumor-ids ID1 ID2`)
		* For example, the tumor reads might come from a sample with the identifier m63080_182826_000329_s2, and a given read might have an ID that looks something like m63080_182826_000329_s2/43058412/ccs. The ID to specify for this example would be m63080_182826_000329_s2.
		* If pooling tumor reads from multiple runs, they can be specified in a comma-delimited list (e.g., m63080_182826_000329_s2,m70605_435710_073329_s4)
	* `--read-sep`: the separator used within read IDs, with the assumption that the text before the first separator will match either a tumor sample ID or a normal sample ID
		* For example, suppose you have reads IDs of the format m63080_182826_000329_s2/43058412/ccs. You would specify / as the `--read-sep`, and m63080_182826_000329_s2 should either be specified in `--tumor-ids` or be the identifier of a normal sample
* Optional arguments
	* `--min-reads`: minimum number of reads a tumor-only node must be supported by to be considered a candidate (default 2)
	* `--min-mapq`: minimum MAPQ a tumor-only node alignment must have to be considered a candidate (default 10)
	* `-t`: number of threads used by minimap2 during alignment (default 3) 

## 3) SV Calling
The last step of the pipeline uses the alignments generated from the preprocessing step and the original co-assembly graph to call somatic breakpoints. The command can be run with

```
./color-sv call -o /output/directory/ --graph coassembly_graph.gfa --filter mask_regions.bed [optional flags]
```

where the directory specified in `-o` must contain an `intermediate_output`directory generated from a [preprocessing step](#preprocess). This command will save the call sets for translocations and all SVs in the output directory as `translocations_region_filtered.sv` and `sv_calls_region_filtered.sv`, respectively. The command will also save versions of the call sets prior to filtering with the `--filter` file as `translocations.sv` and `sv_calls.sv`.

More information about the options:

* Required arguments
	* `--graph`: path to a joint assembly graph in .gfa format (should be the same as the one used in the [preprocessing step](#preprocess))
	* `--filter`: a BED file containing regions to remove from the call set (an example with the CHM13 centromere regions is included in the `examples` directory)
* Optional arguments
	* `-k`: maximum number of layers to search during the breadth-first search for local connectedness (default 10)
		* i.e., a graph will be considered locally disconnected if its neighbors have a distance of more than `k`
	* `-q`: minimum MAPQ of alignments when extracting breakpoints from tumor-only node alignments
	* `-Q`: minimum MAPQ for alignment ends when extracting breakpoints from tumor-only node alignments
	* `--index-bin-size`: number of nodes per chunk when indexing assembly graph file (default 100)
		* affects runtime but not final results
